### ISSUE TITLE
머딥 - 20d (기본 신경망2)

### ISSUE THREAD

**Issue by vmtmxmf5** [vmtmxmf5/Python-Basics#56]
순수하게 오차역전파 구현하려면?
- softmax랑 cross entropy error 클래스 2개를 구현한 뒤에 back prop하면 내가 배웠던 오차역전파 구현할 수 있다!

**Comment by vmtmxmf5** [comment]
# 목차
1. Affine
2. SoftmaxWithLoss


## 1. Affine
- **꼭 기억할 것!** 생성자에서 무엇을 저장해야 하는가? 

| 생성자 저장 |
|--|
| 파라미터 초기값 |
| dL/dw 를 구하기 위한 X |
| X의 shape |
| 파라미터 업데이트 용 dW, db |

- **1차원 배열은 전치해도 1차원**이다. 그러므로 역전파가 안 된다 ㄷㄷㄷ
따라서 2차원으로 늘려서 쓴다

- **평탄화**해줘야 3차원, 4차원 데이터도 내적할 수 있다

- **편향**은 뉴런 별로 **합**을 구해준다 (배치에 대한 정보를 합한다)
그러면 원래 bias의 shape과 같아진다

- 계산 다한 미분값의 형상을 **원본 데이터 shape**으로 바꿔줘야 한다

### Affine == 가중입력치 레이어(그러므로 W와 b를 갖고 있어야 함)
```sage
class Affine:
    def __init__(self, W, b):
        self.W = W
        self.b = b
        
        self.x = None
        self.original_x_shape = None
        
        # 추후에 경사하강법에서 업데이트 하기 위해 만들어 놓아야 함(=학습)
        # dw, db를 알고 있어야 w = w - eta*(dL/dw)로 업데이트 할 수 있음(=학습)
        self.dW = None
        self.db = None
        
    def forward(self, x):
        self.original_x_shape = x.shape
        
        DATA_SIZE = x.shape[0]
        x = x.reshape(DATA_SIZE, -1)

        self.x = x

        out = np.dot(self.x, self.W) + self.b
        return out

    def backward(self, dout):
        dx = np.dot(dout, self.W.T)
        self.dW = np.dot(self.x.T, dout)
        self.db = np.sum(dout, axis=0)
        
        dx = dx.reshape(*self.original_x_shape) # (12 ,3) -> *(12, 3) -> 12, 3으로 언패킹
        return dx
```
- Question 1. dout을 x의 원래 shape으로 바꿔서 활성화함수 계층에 넘겨주어도 문제  없나???
dout이 5차원 배열로 되돌렸다고 가정하자. **sigmoid** 쓰면 **브로드캐스팅**이 일어나기 때문에 계산이 가능하다!
그러므로 미분값이 매번 affine layer를 넘어갈 때 마다 원래 x의 shape으로 바꿔도 된다    

- 주의! forward로 보낸 z(가중입력치)의 차원과 똑같은 미분값(dout) 차원이 들어와야 한다 
(N, 뉴런수) = (N, 뉴런수)


## Softmax + CEE 콤비네이션
사전준비
```
def softmax(x):
    if x.ndim == 2:
        x = x.T
        x = x - np.max(x, axis=0)
        y = np.exp(x) / np.sum(np.exp(x), axis=0)
        return y.T 
    x = x - np.max(x) # 오버플로 대책
    return np.exp(x) / np.sum(np.exp(x))

def cross_entropy_error(y, t):
    if y.ndim == 1:
        t = t.reshape(1, t.size)
        y = y.reshape(1, y.size)
    # 훈련 데이터가 원-핫 벡터라면 정답 레이블의 인덱스로 반환
    if t.size == y.size:
        t = t.argmax(axis=1)
    batch_size = y.shape[0]
    return -np.sum(np.log(y[np.arange(batch_size), t] + 1e-7)) / batch_size
```

### 출력층의 오차?
- backward 시 CEE는 y-t, MSE는 (y-t)h'(z)
- 그러나 **활성화함수(소프트맥스)를 넘어오면 y-t만** 남는다!

- 주의할 점! target의 원핫인코딩 유무
원핫인코딩 되어 있으면 걍 넘기면 되고, 아니면 1 - y[np.arange(batch_size), t]해주면 됨
(t는 정답 클래스 자너)

```sage
class SoftmaxWithLoss:
    def __init__(self):
        self.loss = None
        self.y = None # 역전파 때 사용
        self.t = None # 역전파 때 y-t로 저장해서 사용
        
    def forward(self, x, t):
        self.t = t
        self.y = softmax(x)
        self.loss = cross_entropy_error(self.y, self.t)
        return self.loss
    
    def backward(self, dout=1):
        batch_size = self.t.shape[0]
        
        # y의 원소 개수는 항상 클래스의 개수와 같다
        # t가 원핫인코딩이 되어 있으면 마찬가지로 열개수가 같을 것이고, 안 되어 있으면 1열일 것이다
        if self.t.size == self.y.size:
            dx = (self.y - self.t) / batch_size
            
        else:
            dx = self.y.copy()
            # 오차를 모두 음수로 고려하겠다
            dx[np.arange(batch_size), self.t] -= 1
            dx = dx / batch_size
        
        return dx
```
